### PR TITLE
more encoding flexibility

### PR DESCRIPTION
**Pull request by scottlamb**
* in video toolbox H.265 encoder, allow setting level and max bitrate. Note I originally would automatically cap the max bitrate based on the level, but this might have enough undesirable effects that it's worth being able to play with it in the caller. They can use `H265ProfileLevel::max_bitrate` if desired.

* represent the encoder dropping a frame (altering the trait for all encoders). Video toolbox apparently does this sometimes based on the level and max bitrate selected; I had to track down a confusing `OSStatus(0)` locally to end up here.

* add some context to errors on video encoder creation. It turns out the GitHub Actions CI machine doesn't support setting data rate limits, but my M1 laptop does.

### CODE REVIEW

File: macos/core-foundation/src/string.rs [hunk 1]
```diff
@@ -0,0 +1,29 @@
+use super::*;
+
+#[repr(transparent)]
+pub struct StringRef(sys::CFStringRef);
+
+impl StringRef {
+    pub fn from_static(value: &'static str) -> Self {
+        Self(unsafe {
+            sys::CFStringCreateWithBytesNoCopy(
+                std::ptr::null_mut(),
```
Comment (lzuosym) [inline macos/core-foundation/src/string.rs:10]: same question `null()` should be good.

File: macos/core-foundation/src/array.rs [hunk 2]
```diff
@@ -23,3 +23,19 @@ impl Array {
         }
     }
 }
+
+impl<T: CFType> From<&[T]> for Array {
+    fn from(value: &[T]) -> Self {
+        Self(unsafe { sys::CFArrayCreate(std::ptr::null_mut(), value.as_ptr() as _, value.len() as _, &sys::kCFTypeArrayCallBacks) })
```
Comment (lzuosym) [inline macos/core-foundation/src/array.rs:29]: I think it doesn't need to be `null_mut()`.

File: macos/video-toolbox/build.rs [hunk 3]
```diff
@@ -20,6 +20,7 @@ fn main() {
             .allowlist_var("kVTVideoEncoderSpecification_.+")
             .allowlist_var("kVTCompressionPropertyKey_.+")
             .allowlist_var("kVTEncodeFrameOptionKey_.+")
+            .allowlist_var("kVTProfileLevel_.+")
```
Comment (lzuosym) [inline macos/video-toolbox/build.rs:23]: do we have to expose variables with prefix `kVTProfileLevel_`?

File: macos/video-toolbox/src/video_encoder.rs [hunk 4]
```diff
@@ -1,14 +1,109 @@
...
+            H265ProfileLevel::Main3_0 => "HEVC_Main_3_0",
+            H265ProfileLevel::Main3_1 => "HEVC_Main_3_1",
+            H265ProfileLevel::Main4_0 => "HEVC_Main_4_0",
+            H265ProfileLevel::Main4_1 => "HEVC_Main_4_1",
+            H265ProfileLevel::Main5_0 => "HEVC_Main_5_0",
+            H265ProfileLevel::Main5_1 => "HEVC_Main_5_1",
+            H265ProfileLevel::Main5_2 => "HEVC_Main_5_2",
+            H265ProfileLevel::Main6_0 => "HEVC_Main_6_0",
+            H265ProfileLevel::Main6_1 => "HEVC_Main_6_1",
+            H265ProfileLevel::Main6_2 => "HEVC_Main_6_2",
```
Comment (lzuosym) [inline macos/video-toolbox/src/video_encoder.rs:93]: just curious where do these string values documented for each profile`"HEVC_Main_6_2", "HEVC_Main_6_1"`...?

**Review by lzuosym — CHANGES_REQUESTED** [review]

File: macos/video-toolbox/build.rs (hunk 3 above)
Comment (scottlamb) [reply macos/video-toolbox/build.rs:23]: No. I was trying to use them, but the constants for specific HEVC levels don't exist anyway, so...removed.

File: macos/video-toolbox/src/video_encoder.rs (hunk 4 above)
Comment (scottlamb) [reply macos/video-toolbox/src/video_encoder.rs:93]: They're not. Added a comment about it.

**Review by lzuosym — APPROVED** [review]

File: macos/video-toolbox/build.rs (hunk 3 above)
Comment (lzuosym) [reply macos/video-toolbox/build.rs:23]: It hasn't been removed yet, but there's no harm in keeping it.

**Pull request merged**